### ISSUE TITLE
Hyper-elastic visco-plastic system has broken threading

### ISSUE THREAD

**Issue by dschwen** [idaholab/moose#6233]
As reported [here](https://groups.google.com/forum/#!topic/moose-users/FiTDRnpit74) by Sang Inn Woo.

The HEVP system uses ElementUserObjects from Materials to call `computeValue` directly (basically as a pluggable _Function_ object. The HEVP system relies on material properties being accessible in these user objects. Unfortunately the getUserObject call always returns the thread 0 copy, so all threads of the material class call the thread 0 version of the element user object (which has material properties for thread 0, thus using the wrong values for all material threads except 0). Additionally the thread 0 material properties will be reinitialized everytime the material thread 0 steps to the next element, which can be **right in the middle** of a computation of a different thread (thus the material properties just suddenly change causing the other threads to use undefined values).

**Comment by permcody** [comment]
I'd be curious to see what you are planning to do about this. Just so it's on the record, the state of looped user objects and postprocessors is not defined **during** execution. The whole reason we have `finalize` and `threadJoin` is to consolidate partial information into a single master copy on thread zero and usually processor zero. If we get in the business of handing out the threaded copies of your object then we put the framework in a tough situation. Do we simply copy thread zero's data to all threads? Do we make a different recommendation on what should be implemented in `threadJoin`? Yes, I know that you can make this one case work but "you allow it" will bite us all in the :horse:

**Comment by dschwen** [comment]
@permcody let me answer that question with 9b2aeea. Let's discuss.

**Comment by dschwen** [comment]
I plan on introducing a class `FunctionElementUserObject` (or whatever better name anyone can come up with). The test

```
dynamic_cast<const ElementUserObject *>(&uo) == NULL
```

Should of course be 

```
dynamic_cast<const FunctionElementUserObject *>(&uo) == NULL
```

**Comment by friedmud** [comment]
@permcody has this right.  Looping UserObjects should not be called to use element data... only to retrieve data collapsed in finalize() and threadJoin().

**Comment by dschwen** [comment]
Closed by #6235
